### PR TITLE
Rest - support cross file reference with loop reference

### PR DESCRIPTION
**Pull request by hellosnow**
1. Support external file reference as following:
```json
{
  "in": "body",
  "name": "bodyparam",
  "required": true,
  "schema": {
    "$ref": "./externalRefWithDefinition.json#/definitions/refContainerUrl"
  }
}
```
2. Support to resolve loop reference: A.json reference B.json while B.json reference A.json.

### CODE REVIEW

File: src/Microsoft.DocAsCode.Build.RestApi/Swagger/Internals/JsonLocation/JsonLocationInfo.cs [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+        public override bool Equals(object obj)
+        {
+            var other = obj as JsonLocationInfo;
+            if (other == null)
+            {
+                return false;
+            }
+
+            return string.Equals(FilePath, other.FilePath) && string.Equals(JsonLocation, other.JsonLocation);
```
Comment (vwxyzh) [inline src/Microsoft.DocAsCode.Build.RestApi/Swagger/Internals/JsonLocation/JsonLocationInfo.cs]: >string.Equals [](start = 19, length = 13)

FilePathComparer.OSPlatformSensitiveStringComparer?

File: src/Microsoft.DocAsCode.Build.RestApi/RestApiHelper.cs [hunk 2]
```diff
@@ -45,41 +47,57 @@ public static SwaggerFormattedReference FormatReferenceFullPath(string reference
...
 
-            // Not decode for JSON String Representation
-            if (reference.StartsWith("/"))
+            // External reference
+            if (PathUtility.IsRelativePath(reference) && reference.Contains(JsonExtension))
             {
-                var fileName = reference.Split('/').Last();
-                return new SwaggerFormattedReference
+                // For example "file.json"
+                if (reference.EndsWith(JsonExtension))
```
Comment (superyyrrzz) [inline src/Microsoft.DocAsCode.Build.RestApi/RestApiHelper.cs]: >if (reference.EndsWith(JsonExtension)) [](start = 16, length = 38)

should ignore case?

File: src/Microsoft.DocAsCode.Build.RestApi/Swagger/Internals/JsonLocation/JsonLocationInfo.cs [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+        public override bool Equals(object obj)
+        {
+            var other = obj as JsonLocationInfo;
+            if (other == null)
+            {
+                return false;
+            }
+
+            return FilePathComparer.OSPlatformSensitiveStringComparer.Equals(FilePath, other.FilePath)
+                && FilePathComparer.OSPlatformSensitiveStringComparer.Equals(JsonLocation, other.JsonLocation);
```
Comment (vwxyzh) [inline src/Microsoft.DocAsCode.Build.RestApi/Swagger/Internals/JsonLocation/JsonLocationInfo.cs]: >FilePathComparer.OSPlatformSensitiveStringComparer [](start = 19, length = 50)

always case sensitive for json pointer.

File: src/Microsoft.DocAsCode.Build.RestApi/Swagger/Internals/JsonLocation/JsonLocationInfo.cs [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+        }
+
+        public override int GetHashCode()
+        {
+            return new { FilePath, JsonLocation }.GetHashCode();
```
Comment (vwxyzh) [inline src/Microsoft.DocAsCode.Build.RestApi/Swagger/Internals/JsonLocation/JsonLocationInfo.cs]: also update GetHashCode.

File: src/Microsoft.DocAsCode.Build.RestApi/Swagger/Internals/JsonLocation/JsonLocationInfo.cs (hunk 3 above)
Comment (hellosnow) [inline src/Microsoft.DocAsCode.Build.RestApi/Swagger/Internals/JsonLocation/JsonLocationInfo.cs]: yes you're right, should not ignore case

---
In reply to: [101440772](https://github.com/dotnet/docfx/pull/1308#discussion_r101440772) [](ancestors = 101440772)

File: src/Microsoft.DocAsCode.Build.RestApi/RestApiHelper.cs (hunk 2 above)
Comment (hellosnow) [inline src/Microsoft.DocAsCode.Build.RestApi/RestApiHelper.cs]: just realize the json pointer should be case sensitive.

---
In reply to: [101434772](https://github.com/dotnet/docfx/pull/1308#discussion_r101434772) [](ancestors = 101434772)

Comment (superyyrrzz) [inline src/Microsoft.DocAsCode.Build.RestApi/RestApiHelper.cs]: this condition is to judge whether it's a json, so both json and JSON is allowed?

---
In reply to: [101448000](https://github.com/dotnet/docfx/pull/1308#discussion_r101448000) [](ancestors = 101448000,101434772)

**Comment by vwxyzh** [comment]
:shipit:

File: src/Microsoft.DocAsCode.Build.RestApi/RestApiHelper.cs (hunk 2 above)
Comment (hellosnow) [inline src/Microsoft.DocAsCode.Build.RestApi/RestApiHelper.cs]: this is file path, should ignore case, I updated wrongly in last update.

---
In reply to: [101452529](https://github.com/dotnet/docfx/pull/1308#discussion_r101452529) [](ancestors = 101452529,101448000,101434772)

**Comment by superyyrrzz** [comment]
:shipit:

File: src/Microsoft.DocAsCode.Build.RestApi/RestApiHelper.cs [hunk 5]
```diff
@@ -45,41 +47,61 @@ public static SwaggerFormattedReference FormatReferenceFullPath(string reference
...
+
+                // For example "file.json#/definitions/reference"
+                if (reference.Contains("#"))
+                {
+                    var values = reference.Split('#');
+                    if (values.Length != 2)
+                    {
+                        throw new InvalidOperationException($"Reference path '{reference}' should contain only one '#' character.");
+                    }
+                    var externalFilePath = ParseReferencePath(values[0]).Item1;
```
Comment (vicancy) [inline src/Microsoft.DocAsCode.Build.RestApi/RestApiHelper.cs]: >values[0] [](start = 62, length = 9)

should we check if values[0] ends with .json?

File: src/Microsoft.DocAsCode.Build.RestApi/RestApiHelper.cs [hunk 6]
```diff
@@ -45,41 +47,61 @@ public static SwaggerFormattedReference FormatReferenceFullPath(string reference
...
+                    var parsedReferencePath = ParseReferencePath(values[1].Substring(1));
+                    return new SwaggerFormattedReference
+                    {
+                        Type = SwaggerFormattedReferenceType.ExternalEmbeddedReference,
```
Comment (vicancy) [inline src/Microsoft.DocAsCode.Build.RestApi/RestApiHelper.cs]: >ExternalEmbeddedReference [](start = 61, length = 25)

why not ExternalReference?

File: src/Microsoft.DocAsCode.Build.RestApi/Swagger/Internals/SwaggerJsonBuilder.cs [hunk 7]
```diff
@@ -255,5 +289,16 @@ private static SwaggerObject ResolveSwaggerObject(SwaggerObjectBase swaggerObjec
 
             throw new ArgumentException($"When resolving reference for {nameof(SwaggerReferenceObject)}, only support {nameof(SwaggerObject)} and {nameof(SwaggerReferenceObject)} as parameter.");
         }
+
+        private static string GetDirectory(string path)
+        {
+            var directory = Path.GetDirectoryName(path);
+            if (string.IsNullOrEmpty(directory))
+            {
+                throw new DocfxException($"Directory of swagger file path {path} should not be null or empty.");
```
Comment (vicancy) [inline src/Microsoft.DocAsCode.Build.RestApi/Swagger/Internals/SwaggerJsonBuilder.cs]: >DocfxException [](start = 26, length = 14)

why not allowing swagger file under root folder?

File: src/Microsoft.DocAsCode.Build.RestApi/Swagger/Internals/SwaggerJsonBuilder.cs [hunk 8]
```diff
@@ -189,21 +212,32 @@ private SwaggerObjectBase ResolveReferences(SwaggerObjectBase swaggerBase, Stack
...
                                 var loopRef = new SwaggerLoopReferenceObject();
                                 loopRef.Dictionary.Add(InternalLoopRefNameKey, new SwaggerValue { Token = swagger.ReferenceName });
                                 return loopRef;
                             }
 
                             // Clone to avoid change the reference object in _documentObjectCache
-                            refStack.Push(referencedObjectBase.Location);
-                            var resolved = ResolveReferences(referencedObjectBase.Clone(), refStack);
+                            refStack.Push(jsonLocationInfo);
+                            var resolved = ResolveReferences(referencedObjectBase.Clone(), jsonLocationInfo.FilePath, refStack); // TODO
```
Comment (hellosnow) [inline src/Microsoft.DocAsCode.Build.RestApi/Swagger/Internals/SwaggerJsonBuilder.cs]: > // TODO [](start = 128, length = 8)

remove the todo...

File: src/Microsoft.DocAsCode.Build.RestApi/Swagger/Internals/SwaggerJsonBuilder.cs (hunk 7 above)
Comment (hellosnow) [inline src/Microsoft.DocAsCode.Build.RestApi/Swagger/Internals/SwaggerJsonBuilder.cs]: "C:\s.json" is supported while "s.json" without directoty is not support. the path is passed by "Path.Combine(file.BaseDir, file.File)" in RestApiDocumentProcessor, so we assume that the directory exists here?

---
In reply to: [101675921](https://github.com/dotnet/docfx/pull/1308#discussion_r101675921) [](ancestors = 101675921)

File: src/Microsoft.DocAsCode.Build.RestApi/RestApiHelper.cs (hunk 5 above)
Comment (hellosnow) [inline src/Microsoft.DocAsCode.Build.RestApi/RestApiHelper.cs]: yes better check, I'll add.

---
In reply to: [101673560](https://github.com/dotnet/docfx/pull/1308#discussion_r101673560) [](ancestors = 101673560)

File: src/Microsoft.DocAsCode.Build.RestApi/Swagger/Internals/SwaggerJsonBuilder.cs (hunk 7 above)
Comment (vicancy) [inline src/Microsoft.DocAsCode.Build.RestApi/Swagger/Internals/SwaggerJsonBuilder.cs]: is it better to not highly dependent on the caller?

---
In reply to: [101679374](https://github.com/dotnet/docfx/pull/1308#discussion_r101679374) [](ancestors = 101679374,101675921)

Comment (hellosnow) [inline src/Microsoft.DocAsCode.Build.RestApi/Swagger/Internals/SwaggerJsonBuilder.cs]: not depend on the caller, but need the directory to combine with external path's relative path.

---
In reply to: [101680379](https://github.com/dotnet/docfx/pull/1308#discussion_r101680379) [](ancestors = 101680379,101679374,101675921)

Comment (hellosnow) [inline src/Microsoft.DocAsCode.Build.RestApi/Swagger/Internals/SwaggerJsonBuilder.cs]: offline synced, relative path should be supported, so will remove the check.

---
In reply to: [101680622](https://github.com/dotnet/docfx/pull/1308#discussion_r101680622) [](ancestors = 101680622,101680379,101679374,101675921)

**Comment by vicancy** [comment]
:shipit:

**Pull request merged**